RABID-2 Add Netlink support to luasocket

marcelorcs opened the pull request:
Extended the luasocket API to be compatible with netlink sockets

marcelorcs committed:
Add support for netlink

marcelorcs committed:
Add connect() to netlink sockets

iru- commented on src/makefile:
```diff
@@ -18,7 +18,7 @@ PLAT?=linux
 # LUAV: 5.1 5.2
 # lua version to build against
-LUAV?=5.1
+LUAV?=5.3
```
Why do we change that? We could have a global `LUAV=5.3` variable for our build instead.

marcelorcs replied:
I'm not too sure how to do that. What exactly do you mean?

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+/*=========================================================================*\
+* Unix domain socket dgram submodule
```
This comment is misleading, not the unix domain socket submodule, is it?

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+
+static const char *netlink_trybind(p_netlink nl,int grp);
+
+/* unixdgram object methods */
```
Misleading again.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+* Send data through connected netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_send(lua_State *L){
+	p_netlink nl= (p_netlink) auxiliar_checkclass(L,"netlink{connected}",1);
```
Use spaces on both sides of `=` and after `,`.

LeoLAR replied:
Change checkclass so as to not generate an error.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+static int meth_send(lua_State *L){
+	p_netlink nl= (p_netlink) auxiliar_checkclass(L,"netlink{connected}",1);
+	size_t payload_size, sent = 0;
+	const char *payload=luaL_checklstring(L,2,&payload_size);
```
Use spaces on both sides of `=` and after `,`.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+	const char *payload=luaL_checklstring(L,2,&payload_size);
+	p_timeout tm = &nl->tm; 
+
+	int flags = luaL_optinteger(L,3,0);
```
Need a space after `,`.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+
+	memcpy(NLMSG_DATA(nlh),payload,payload_size);
+
+	int err = socket_send(&nl->fd,(char *)nlh,nlh->nlmsg_len,&sent,tm);
```
spaces again

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+	memcpy(NLMSG_DATA(nlh),payload,payload_size);
+
+	int err = socket_send(&nl->fd,(char *)nlh,nlh->nlmsg_len,&sent,tm);
+	if(err!=IO_DONE){
```
and again

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+	int err = socket_send(&nl->fd,(char *)nlh,nlh->nlmsg_len,&sent,tm);
+	if(err!=IO_DONE){
+		lua_pushnil(L);
+		lua_pushliteral(L,"error sending message");
```
I'll stop commenting on spaces, but you should check all of them

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+	int err= socket_recv(&nl->fd,(char *)nlh,nlh->nlmsg_len,&got,tm);
+    	if (err != IO_DONE && err != IO_CLOSED) {
+        	lua_pushnil(L);
+       		lua_pushstring(L,socket_strerror(err));
```
Wrong indentation level.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+
+	struct nlmsghdr *nlh;
+
+	nlh = (struct nlmsghdr *)malloc(NLMSG_SPACE(payload_size));
```
Is there any reason for not checking your malloc's returns?

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+
+	memcpy(NLMSG_DATA(nlh),payload,payload_size);
+
+	int err = socket_sendto(&nl->fd,(char *)nlh,nlh->nlmsg_len,&sent,(struct sockaddr *)&addr,sizeof(addr),tm);
```
Line bigger than 80 chars, split it and check for other occurrences.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+/*-------------------------------------------------------------------------*\
+* Receives data from a netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_receive(lua_State *L){
```
Space missing before the opening `{`

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+    return 1;
+}
+
+/* this is very dangerous, but can be handy for those that are brave enough */
```
Is this is dangerous, why are we adding it here?

renatomaia replied:
Actually it is dangerous, but it is also handy in some situations. For example, we can use this to create make a socket in Lua to refer a socket created by 'libwebsockets' for example. Since LuaSocket does not provide a C API to do this, this sort of fallback becames handy.

We should add this anyway just to be compliant with what LuaSocket does currently, since we plan to contribute this code back to LuaSocket official repository.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+/*-------------------------------------------------------------------------*\
+* Binds an object to an address
+\*-------------------------------------------------------------------------*/
+static const char *netlink_trybind(p_netlink nl,int grp) {
```
Are we sure `nl` is never null? If not, we should check it.

marcelorcs replied:
meth_bind() (which is the function that calls trybind) already checks that through the auxiliar.h function auxiliar_checkclass()

iru- replied:
Also missing space

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+	addr.nl_pid=nl->srcpid;
+	getpeername(nl->fd,(struct sockaddr *)&addr,(socklen_t *)sizeof(addr));
+
+	//int res=bind(p,(struct sockaddr *)&addr,sizeof(addr));
```
Why is this comment here?

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+	return socket_strerror(err);
+}
+
+static int meth_bind(lua_State *L)
```
Use a single opening brace style in the whole file.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+    	p_netlink nl = (p_netlink) auxiliar_checkclass(L, "netlink{unconnected}", 1);
+	int pid=luaL_checkinteger(L,2);
+	int grp=luaL_optinteger(L,3,0);
+    	if(!nl){
```
`nl != NULL`, since nl is not a boolean.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+		socket_connect(&nl->fd,(struct sockaddr *)&addr,(socklen_t)sizeof(addr),&nl->tm);
+		auxiliar_setclass(L,"netlink{unconnected}",1);
+		return 0;
+	}
```
Add an empty line

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+	if (getpeername(nl->fd, (SA *) &peer, &peer_len) < 0) {
+		lua_pushnil(L);
+		lua_pushstring(L, socket_strerror(errno));
+	    	return 2;
```
Use an `else` clause here instead of such closely spaced equal returns.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+{
+    t_socket sock;
+    int err = socket_create(&sock, AF_NETLINK, SOCK_RAW, NETLINK_USERSOCK);
+    /* try to allocate a system socket */
```
Too many comments, strive for code so clean that comments aren't needed. This comments actually disturb the reading.

marcelorcs replied:
If it was 100% my API I'd agree with you, but I'm following the standard set by Diego, so I can hopefully send him a pull request eventually

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+    int err = socket_create(&sock, AF_NETLINK, SOCK_RAW, NETLINK_USERSOCK);
+    /* try to allocate a system socket */
+    if (err == IO_DONE){
+        /* allocate unixdgram object */
```
Wrong, not a unixdgram.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+        /* initialize remaining structure fields */
+        socket_setnonblocking(&sock);
+        nl->fd = sock;
+	nl->type=SOCK_RAW;
```
Wrong indentation

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+        socket_setnonblocking(&sock);
+        nl->fd = sock;
+	nl->type=SOCK_RAW;
+        /*io_init(&un->io, (p_send) socket_send, (p_recv) socket_recv,
```
Why is this comment here?

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+        timeout_init(&nl->tm, -1, -1);
+        /*buffer_init(&un->buf, &un->io, &un->tm);*/
+        return 1;
+    } else {
```
Either use an `else` or a `return` inside the `if`, never both.

iru- commented on src/netlink.h:
```diff
@@ -0,0 +1,24 @@
+
+
+/*=========================================================================*\
+* .h definitions here because i'm lazy lmao 
```
Commenting doesn't make it any more acceptable to write poor code.

iru- commented on src/netlink.h:
```diff
@@ -0,0 +1,24 @@
+#define t_type int
+
+typedef struct t_netlink_{
+	t_socket fd; /*socket number (int)*/
```
You do not seem that lazy with all these useless comments, uh?

iru- commented on src/netlink.h:
```diff
@@ -0,0 +1,24 @@
+	t_pid srcpid; /*port id (int) */
+	t_type type; /*socket type (int)*/
+}t_netlink;
+typedef t_netlink *p_netlink;
```
Why do we hide the fact this is a pointer by typedefing over it?

iru- commented on src/netlink.h:
```diff
@@ -0,0 +1,24 @@
+}t_netlink;
+typedef t_netlink *p_netlink;
+
+int netlink_open(lua_State *L);
```
Why is there a prototype here?

marcelorcs replied:
luasocket.c needs to be able to see that function in order to open the netlink module. All other modules' .h's include that prototype

iru- commented on src/netlink.h:
```diff
@@ -0,0 +1,24 @@
+#define t_groups int
+#define t_type int
+
+typedef struct t_netlink_{
```
Do you use `struct t_netlink_` somewhere? If not, the struct doesn't need a name.

LeoLAR commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+		return 2;
+	}
+	auxiliar_setclass(L,"netlink{connected}",1);
+	lua_pushinteger(L,err);
```
return true instead of err

LeoLAR commented on src/netlink.c:
```diff
@@ -0,0 +1,362 @@
+	        lua_pushstring(L, err);
+	        return 2;
+	}
+        lua_pushnumber(L, 1);
```
return true instead of 1

lneto commented on src/luasocket.c:
```diff
@@ -30,6 +30,7 @@
 #include "inet.h"
 #include "tcp.h"
 #include "udp.h"
+#include "netlink.h"
```
ifdef

marcelorcs replied:
I assume you meant for me to include the ifdef inside netlink.h (which I did), however, since theres no change in the line you tagged, your comment won't disappear.

lneto reviewed (changes requested):
@marcelorcs please, don't do commits like this "fix typos" use fixups and read our workflow. Also, commit the skeleton copied from luasocket as the original author so we can have the visibility of what is yours and what's not.

diegonehab committed:
Add original code skeleton

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+static int meth_setpeername(lua_State *L);
+static int meth_getsockpid(lua_State *L);
+
+static const char *netlink_trybind(p_netlink nl,int grp);
```
space after comma

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+
+	int flags = luaL_optinteger(L, 3, 0);
+
+	struct nlmsghdr *nlh;
```
if you are declaring and initializing, like the lines above, why do differently with `nlh`? be consistent in style.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	struct nlmsghdr *nlh;
+
+	nlh = (struct nlmsghdr *)malloc(NLMSG_SPACE(payload_size));
+	if(nlh==NULL) {
```
missing spaces around `==` and after `if`

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	nlh = (struct nlmsghdr *)malloc(NLMSG_SPACE(payload_size));
+	if(nlh==NULL) {
+		lua_pushnil(L);
+		lua_pushliteral(L,"error allocating memory in send()");
```
spaces after comma

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	memcpy(NLMSG_DATA(nlh), payload, payload_size);
+
+	int err = socket_send(&nl->fd, (char *)nlh, nlh->nlmsg_len, &sent, tm);
+	if(err!=IO_DONE){
```
spaces after `if` and `)`, between `!=`

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+* Send data through unconnected netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_sendto(lua_State *L) {
+	p_netlink nl= (p_netlink) auxiliar_checkclass(L, "netlink{unconnected}",1);
```
space before `=`, after comma. Also, be consistent in the spaces after a type cast, the malloc one above didn't have one, this one does.

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	addr.nl_groups = groups;
+
+	nlh = (struct nlmsghdr *)malloc(NLMSG_SPACE(payload_size));
+	if(nlh==NULL) {
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	nlh = (struct nlmsghdr *)malloc(NLMSG_SPACE(payload_size));
+	if(nlh==NULL) {
+		lua_pushnil(L);
+		lua_pushliteral(L,"error allocating memory in send()");
```
space after comma

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+		lua_pushnil(L);
+		lua_pushliteral(L,"error allocating memory in send()");
+		return 2;
+	}
```
add a newline

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	nlh->nlmsg_len = NLMSG_SPACE(payload_size);
+	nlh->nlmsg_pid = nl->srcpid;
+	nlh->nlmsg_flags = flags;
+
```
remove the newline

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+
+	int err = socket_sendto(&nl->fd, (char *)nlh, nlh->nlmsg_len, &sent, 
+			(struct sockaddr *)&addr,sizeof(addr), tm);
+
```
remove the blank line

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	int err = socket_sendto(&nl->fd, (char *)nlh, nlh->nlmsg_len, &sent, 
+			(struct sockaddr *)&addr,sizeof(addr), tm);
+
+	if(err!=IO_DONE) {
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+* Receives data from a netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_receive(lua_State *L) {
+	p_netlink nl = (p_netlink) auxiliar_checkclass(L, "netlink{connected}", 1);
```
cast spacing

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	size_t got;
+	p_timeout tm = &nl->tm;
+
+	nlh=(struct nlmsghdr *)malloc(NLMSG_SPACE(MAX_PAYLOAD));
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	p_timeout tm = &nl->tm;
+
+	nlh=(struct nlmsghdr *)malloc(NLMSG_SPACE(MAX_PAYLOAD));
+	if(nlh==NULL) {
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	nlh=(struct nlmsghdr *)malloc(NLMSG_SPACE(MAX_PAYLOAD));
+	if(nlh==NULL) {
+		lua_pushnil(L);
+		lua_pushliteral(L,"error allocating memory in send()");
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+		return 2;
+	}
+	memset(nlh, 0, NLMSG_SPACE(MAX_PAYLOAD));
+	nlh->nlmsg_len=NLMSG_SPACE(MAX_PAYLOAD);
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	memset(nlh, 0, NLMSG_SPACE(MAX_PAYLOAD));
+	nlh->nlmsg_len=NLMSG_SPACE(MAX_PAYLOAD);
+
+	int err= socket_recv(&nl->fd, (char *)nlh, nlh->nlmsg_len, &got, tm);
```
space

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	int err= socket_recv(&nl->fd, (char *)nlh, nlh->nlmsg_len, &got, tm);
+    	if (err != IO_DONE && err != IO_CLOSED) {
+        	lua_pushnil(L);
+       		lua_pushstring(L, socket_strerror(err));
```
wrong indentation level

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	size_t got;
+	p_timeout tm = &nl->tm;
+
+	nlh=(struct nlmsghdr *)malloc(NLMSG_SPACE(MAX_PAYLOAD));
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	p_timeout tm = &nl->tm;
+
+	nlh=(struct nlmsghdr *)malloc(NLMSG_SPACE(MAX_PAYLOAD));
+	if(nlh==NULL) {
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	nlh=(struct nlmsghdr *)malloc(NLMSG_SPACE(MAX_PAYLOAD));
+	if(nlh==NULL) {
+		lua_pushnil(L);
+		lua_pushliteral(L,"error allocating memory in send()");
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+		lua_pushliteral(L,"error allocating memory in send()");
+		return 2;
+	}
+	memset(nlh,0,NLMSG_SPACE(MAX_PAYLOAD));
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	socklen_t len = sizeof(dst);
+	int err = socket_recvfrom(&nl->fd, (char *)nlh, nlh->nlmsg_len, &got, 
+			(struct sockaddr *)&dst, &len, tm);
+
```
remove blank line

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+
+    	if (err != IO_DONE && err != IO_CLOSED) {
+        	lua_pushnil(L);
+       		lua_pushstring(L, socket_strerror(err));
```
wrong indentation level

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+\*-------------------------------------------------------------------------*/
+static int meth_getfd(lua_State *L) {
+	p_netlink nl = (p_netlink) auxiliar_checkgroup(L, "netlink{any}", 1);
+    	lua_pushnumber(L, (int) nl->fd);
```
cast space

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+* Select support methods
+\*-------------------------------------------------------------------------*/
+static int meth_getfd(lua_State *L) {
+	p_netlink nl = (p_netlink) auxiliar_checkgroup(L, "netlink{any}", 1);
```
cast space

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+* Receives data from a netlink socket
+\*-------------------------------------------------------------------------*/
+static int meth_receivefrom(lua_State *L) {
+	p_netlink nl = (p_netlink) auxiliar_checkclass(L, "netlink{unconnected}", 1);
```
cast space

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+}
+
+static int meth_getsockpid(lua_State *L) {
+	p_netlink nl = auxiliar_checkgroup(L, "netlink{any}", 1);
```
missing cast to `p_netlink`

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+
+static int meth_bind(lua_State *L) {
+    	p_netlink nl = (p_netlink) auxiliar_checkclass(L, "netlink{unconnected}", 1);
+	int pid=luaL_checkinteger(L, 2);
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+static int meth_bind(lua_State *L) {
+    	p_netlink nl = (p_netlink) auxiliar_checkclass(L, "netlink{unconnected}", 1);
+	int pid=luaL_checkinteger(L, 2);
+	int grp=luaL_optinteger(L, 3, 0);
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+    	p_netlink nl = (p_netlink) auxiliar_checkclass(L, "netlink{unconnected}", 1);
+	int pid=luaL_checkinteger(L, 2);
+	int grp=luaL_optinteger(L, 3, 0);
+    	if(nl != NULL) {
```
space

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	p_netlink nl;
+	struct sockaddr_nl addr;
+
+	if(lua_isnone(L, 2)) {
```
space

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	struct sockaddr_nl addr;
+
+	if(lua_isnone(L, 2)) {
+		nl=auxiliar_checkclass(L, "netlink{connected}", 1);
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+		return 0;
+	}
+
+	nl=auxiliar_checkclass(L, "netlink{unconnected}", 1);
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	addr.nl_groups = grps;
+	addr.nl_family = AF_NETLINK;
+
+	int err=socket_connect(&nl->fd, (struct sockaddr *)&addr, 
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+
+	int err=socket_connect(&nl->fd, (struct sockaddr *)&addr, 
+			(socklen_t)sizeof(addr), &nl->tm);
+
```
remove blank line

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+	int err=socket_connect(&nl->fd, (struct sockaddr *)&addr, 
+			(socklen_t)sizeof(addr), &nl->tm);
+
+	if(err!=IO_DONE) {
```
spaces

iru- commented on src/netlink.c:
```diff
@@ -0,0 +1,387 @@
+    		timeout_init(&nl->tm, -1, -1);
+        	return 1;
+	} 
+       	lua_pushnil(L);
```
wrong indentation

iru- commented on src/netlink.h:
```diff
@@ -0,0 +1,37 @@
+	t_timeout tm;  
+	t_pid srcpid; 
+	t_type type; 
+}t_netlink;
```
space

lneto commented on src/luasocket.c:
```diff
@@ -30,6 +30,9 @@
 #include "inet.h"
 #include "tcp.h"
 #include "udp.h"
+#ifndef NETLINK_H
```
it's not enough to put ifdef only here.. you should apply it to all code that depends on netlink. also, I don't think NETLINK_H is right here.. you should use convention used by luasocket to other platform-dependent code..

lneto reviewed (changes requested):
@marcelorcs review this commit https://github.com/getCUJO/luasocket/pull/1/commits/008e6eed74579f8a09d104cc41c2d10afbf42035

please, don't remove original codes without needing.. e.g., line breaks, section descriptors, nor change original format like you did on meth_receive

lneto commented on src/netlink.h:
```diff
@@ -0,0 +1,37 @@
+
+#define t_pid int
+#define t_groups int
+#define t_type int
```
why defining this?

marcelorcs replied:
Following the style used by Diego. But actually, it's supposed to be a typedef. Theres a similar section of code in usocket.h

lneto commented on src/netlink.h:
```diff
@@ -0,0 +1,37 @@
+    t_pid srcpid; 
+    t_type type; 
+} t_netlink;
+typedef t_netlink *p_netlink;
```
why defining this?

marcelorcs replied:
Following the style used by Diego

lneto commented on src/netlink.c:
```diff
@@ -0,0 +1,385 @@
+\*-------------------------------------------------------------------------*/
+static int meth_receive(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{connected}", 1);
+    struct nlmsghdr *nlh = (struct nlmsghdr *)malloc(NLMSG_SPACE(MAX_PAYLOAD));
```
why using malloc?

marcelorcs replied:
I need to allocate a sufficiently large buffer to store the message. Do you suggest using something else?

lneto commented on src/netlink.c:
```diff
@@ -0,0 +1,385 @@
+    nlh->nlmsg_flags = flags;
+    memcpy(NLMSG_DATA(nlh), payload, payload_size);
+
+    err = socket_send(&nl->fd, (char *)nlh, nlh->nlmsg_len, &sent, tm);
```
you are not marking timeout here like the original code does.. is this intentional?

marcelorcs replied:
It is not and I will proceed to re-add that

marcelstanley reviewed (changes requested):
Please correct me if I'm wrong but you should not "fixup" imported code. Other than that, you should review your commits again and follow our development workflow strictly, you're mixing up where you should use "fixups" and where you should not. Commits like _fix missing code from netlink.c_ and _fix code based on PR reviews_ should not exist

lneto reviewed (changes requested):
please squash your commits.. I know I said that it's alright.. but there are too many levels now.. it will make it easier to review.. thanks..

LeoLAR commented on src/netlink.h:
```diff
@@ -1,24 +1,37 @@
 
+typedef t_pid int;
+typedef t_groups int;
+typedef t_type int;
```
Shouldn't it be "typedef int t_type"?

marcelorcs replied:
yes it should

marcelorcs commented:
All currently known issues fixed

renatomaia commented on src/luasocket.c:
```diff
@@ -51,6 +54,9 @@ static const luaL_Reg mod[] = {
     {"select", select_open},
+#ifdef LUASOCKET_NETLINK
+    {"netlink", netlink_open},
+#endif
```
Why did you included the Netlink into the `socket.core` module instead of providing it as a separate module like `socket.unix` which was the one you were using as template?

marcelorcs replied:
It's because the netlink socket is only slightly different than a udp socket, so it seemed fair to not separate them.

renatomaia replied:
Even if being similar or different were a good reason to include in the core part of LuaSocket (which is not!) then UNIX domain socket would have to be included as well because it is much more similar to "inet" sockets than Netlink. So you either adopt the same approach for 'socket.unix' or you change the way you added your Netlink support.

renatomaia commented on src/netlink.c:
```diff
@@ -0,0 +1,365 @@
+    size_t payload_size;
+    const char *payload = luaL_checklstring(L, 2, &payload_size);
+    int flags = luaL_optinteger(L, 3, 0);
+    struct nlmsghdr nlh[NLMSG_SPACE(payload_size)];
```
This allocates `nlh` with `NLMSG_SPACE(payload_size) * sizeof(struct nlmsghdr)` bytes, which is much more than the `sizeof(struct nlmsghdr)` bytes you actually need.

marcelorcs replied:
I need to allocate enough bytes to nlh, so that I can fit the payload in it. Therefore I'm using the macro NLMSG_SPACE(), that gives me the total size of the message (header + payload). So, a single sizeof() won't do the trick. However, doing it that way is still more. The correct way I believe would be something like this:
```C
char buff[NLMSG_SPACE(payload_size)];
struct nlmsghdr *nlh = (struct nlmsghdr *)buff;
```
or just a malloc, since payload_size is not known in compilation time 
```C
malloc(NLMSG_SPACE(payload_size));
```

renatomaia commented on src/netlink.c:
```diff
@@ -0,0 +1,365 @@
+    memset(nlh, 0, NLMSG_SPACE(payload_size));
+    nlh->nlmsg_len = NLMSG_SPACE(payload_size);
+    nlh->nlmsg_pid = nl->srcpid;
+    nlh->nlmsg_flags = flags;
```
Which values for flags do you imagine the application will be able to provide? The application programmer will have to know the numeric value of `NLM_F_ACK` for example? Moreover, some values, like the value of `NLM_F_MULTI`, will render the message illegal, doesn't it? This doesn't look like a good API for `netlink:send`.

renatomaia replied:
How about other header fields like `nlmsg_seq` and `nlmsg_type`? The application cannot use it? We're not using it in our protocol over Netlink to talk to out code in the kernel?

marcelorcs replied:
He would have to find out the numeric values, however it's an optional field. If he wanted to specify a flag he could, but it's not necessary. 
Why would that flag render the message illegal? 
nlmsg_seq and nlmsg_type are opaque to the netlink core. We could use them, but they would only serve a purpose if we assigned them a purpose on our implementation.

renatomaia replied:
Why are you telling me exactly the problems I've asked you to address?
Did you bother to read the docs about `NLM_F_MULTI`?
Again, why are you telling me what `nlmsg_seq` and `nlmsg_type` are instead of addressing my point about the application and our code in the kernel not being able to use it?

renatomaia commented on src/netlink.c:
```diff
@@ -0,0 +1,365 @@
+    nlh->nlmsg_pid = nl->srcpid;
+    nlh->nlmsg_flags = flags;
+    memcpy(NLMSG_DATA(nlh), payload, payload_size);
+    timeout_markstart(tm);
```
Isn't all the code above common with the `send` implementation? Shouldn't it be extracted to a common function somehow? So we can fix all the bugs in one place instead of fixing multiples copies of the same buggy code.

marcelorcs replied:
Well yes, but I wanted to stick to the code style as much as I could. Since that's not done much, I didn't do it either, as much as I agree with you.

renatomaia replied:
Don't you think there is a limit to what extent you can use that excuse?

renatomaia commented on src/netlink.c:
```diff
@@ -0,0 +1,365 @@
+\*-------------------------------------------------------------------------*/
+static int meth_receive(lua_State *L) {
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{connected}", 1);
+    struct nlmsghdr nlh[NLMSG_SPACE(MAX_PAYLOAD)];
```
This again allocates much more memory than necessary.

renatomaia commented on src/netlink.c:
```diff
@@ -0,0 +1,365 @@
+    int err;
+
+    memset(nlh, 0, NLMSG_SPACE(MAX_PAYLOAD));
+    nlh->nlmsg_len = NLMSG_SPACE(MAX_PAYLOAD);
```
Why this `MAX_PAYLOAD` limit? Why not let the application provide the size it expects the message to be like in other Lua sockets, like `sock:receive(2048)`.

marcelorcs replied:
That size is not what the application expects to receive, it's just an arbitrary value, that's sufficiently large to store any message that may come. Much like the 1024 size for DNS messages in the appliance code.

renatomaia replied:
Who ever said `MAX_PAYLOAD` is what the application expects to receive? Would you care to enlighten me by pointing me the documentation or whatever thing from which you took this nonsense about 1024 and DNS?

marcelorcs replied:
As far as I understand, you did. Here's what I understood from your sentence: Why instead of defining a ceiling value for the message length, thats about to be received, you send that value instead? And about that DNS size, it's not nonsense. If I got anything wrong about that, it was the actual exact size. I'm not sure you remember when me, @VictorNogueiraRio and @LeoLAR found a bug where DNS messages were not being processed, and instead an error was being thrown that they were too large, but @marcelstanley probably does, since he asked us then, if we could reproduce it. It was some sort of sanity check I believe.

renatomaia commented on src/netlink.c:
```diff
@@ -0,0 +1,365 @@
+    memset(nlh, 0, NLMSG_SPACE(MAX_PAYLOAD));
+    nlh->nlmsg_len = NLMSG_SPACE(MAX_PAYLOAD);
+    socklen_t len = sizeof(dst);
+    timeout_markstart(tm);
```
Again, repeated code, repeated problems... https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

renatomaia commented on src/netlink.c:
```diff
@@ -0,0 +1,365 @@
+    memset(nlh, 0, NLMSG_SPACE(MAX_PAYLOAD));
+    nlh->nlmsg_len = NLMSG_SPACE(MAX_PAYLOAD);
+    timeout_markstart(tm);
+    err = socket_recv(&nl->fd, (char *)nlh, nlh->nlmsg_len, &got, tm);
```
How about multipart messages (`NLM_F_MULTI`), acknowlegment requests (`NLM_F_ACK`) and other flag options? They are all ignored and potentially discards data of the multipart messages?
Why ignore field `nlmsg_pid`? How about `nlmsg_type` and `nlmsg_seq` again?

marcelorcs replied:
Those flags aren't treated here. The flags are read by the netlink core (inside recv, recvfrom and recvmsg), that's why they're specified  only when we send the message. Again, those fields are opaque to netlink core, and only carry any meaning if we arbitrarily add meaning to them.

renatomaia replied:
Again, would you care to enlighten me by pointing me from where did you get this about `recv`, `recvfrom`, and `recvmsg` processing those flags? Because from everything I've read about Netlink that's complete rubbish.

marcelorcs replied:
It is in fact wrong, I actually misinterpreted a couple of lines from Kernel Korner's arcticle. Also, I believe that is done is kernel not in userspace. This is in 'man 7 netlink' for example:

> For reliable transfer the
       sender can request an acknowledgement from the receiver by setting
       the NLM_F_ACK flag.  An acknowledgment is an NLMSG_ERROR packet with
       the error field set to 0.  The application must generate acknowledge‐
       ments for received messages itself.  The kernel tries to send an
       NLMSG_ERROR message for every failed packet.  A user process should
       follow this convention too.

renatomaia commented on src/netlink.c:
```diff
@@ -0,0 +1,365 @@
+    getpeername(nl->fd, (SA *)&addr, (socklen_t *)sizeof(addr));
+    err = socket_bind(&nl->fd, (SA *)&addr, sizeof(addr));
+    return socket_strerror(err);
+}
```
Why did you separate this code in another function? It is only used in one place and it is no more complex than most of other functions.

marcelorcs replied:
I followed the style used by Diego. He uses a try_ function for both bind and connect, so I did the same, except there's no connect here.

renatomaia replied:
So, if Diego's code jumps off a bridge... ah, never mind.

renatomaia commented on src/netlink.c:
```diff
@@ -0,0 +1,365 @@
+                (socklen_t)sizeof(addr), &nl->tm);
+        auxiliar_setclass(L, "netlink{unconnected}", 1);
+        return 0;
+    }
```
Where did this came from? Do you have links for documentation that show this is valid and what it means exactly?

marcelorcs replied:
It came from Diego's mind pretty much. It's an abstraction of udp sockets, since, by definition, they 'cannot' be in a connected or unconnected state. Connecting them only means that the destination port id is stored internally, so that the user can use the send function without having to specify the destination port everytime. Although, theoretically, I believe it's not the most correct thing.

renatomaia replied:
I'm really not interested in your mind reading powers, as much impressive they might seem, so how about you pointing me the original Diego's code that you based on to write this, and more importantly which documentation about Netlink made you think it is correct to pass a `AF_UNSPEC` address to `connect` and what does it mean.

marcelorcs replied:
If you wanted to ask about AF_UNSPEC, perhaps you should've added your comment on line 284 instead of line 289, because even though I can read minds, I guess I can't read yours. You asked "where did this come from" and pointed to a closing curly bracket...The only section of code you could be asking about that showed in the PR comment section was the line with "netlink{unconnected}", so that's what I explained to you.. I only ask you to not be so condescending next time, it's unhealthy (Wouldn't it be better if you just replied, "I know what that means, I was asking about AF_UNSPEC"?). Now to answer your question, I took that from the man page of 'connect'
> Generally, connection-based protocol sockets may successfully connect()
       only once; connectionless protocol sockets may use  connect()  multiple
       times to change their association.  Connectionless sockets may dissolve
       the association by connecting to an address with the  sa_family  member
       of sockaddr set to AF_UNSPEC (supported on Linux since kernel 2.2).

marcelorcs committed:
Move netlink from socket.core to socket.netlink

*Add flag processing function
*Add auxiliar send and receive functions to remove duplicate code

cj-carlos-carvalho commented:
2018 PR, we can re-open it another time if needed.
